uninitialized check

kroening opened the pull request:
Uninitialized local or dynamically allocated variables have an indeterminate initial value.

Reading an indeterminate value _may_ be undefined behavior, or may yield an unspecific value.

This adds a check for uninitialized local variables.  The check is not added as a standard check.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

rod-chapman commented:
Good to see this PR coming. What is your soundness argument? Does this check always detect all cases of uninitialized variables? Are there cases where it is known to miss? What are they?

tautschnig committed:
Initialize nondet locals in regression tests for --uninitialized-check

Replace bare uninitialized local variable declarations with explicit
initialization in ~175 regression test files. These tests use
uninitialized locals as a shorthand for nondeterministic values
(e.g., int x; __CPROVER_assume(x > 0);). This is valid C but
triggers --uninitialized-check when enabled globally.

Prepares for --uninitialized-check becoming a default check.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add comprehensive test suite for --uninitialized-check

28 test cases covering scalar locals, struct/union/array members,
nested structs, per-element array tracking, symbolic indices,
heap allocation, pointer aliasing, goto across initialization,
and various control flow patterns.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

kroening committed:
Implement --uninitialized-check in goto_check_c

Shadow boolean tracking for detecting reads of uninitialized local
variables (C11 §6.3.2.1p2 undefined behavior). The standard scopes
UB to objects that "could have been declared with the register
storage class (never had its address taken)" — address-taken locals
yield merely indeterminate values, not UB.

Design: for each tracked local variable, a shadow boolean $init is
added to the symbol table. The first pass inserts DECL + init=false
after each tracked variable's DECL. The assignment handler sets
init=true on writes, and check_rec asserts init==true on reads.

Features:
- Per-member tracking for struct components (recursive, arbitrary depth)
- Per-element tracking for constant-size arrays (up to 64 elements)
- Symbolic array index handling (conjunction of all element flags)
- Heap allocation tracking via pointer init flags
- Heap aliasing via shared flag symbols
- Union/byte_extract write handling
- Function call return value tracking
- DECL+ASSIGN optimization (skip variables initialized at declaration)
- Skip address-taken (dirty) variables (not UB per C11 §6.3.2.1p2)
- Skip compiler-generated temporaries

Unifies goto-instrument --uninitialized-check to use the same
goto_check_c implementation, removing the old uninitializedt.

Includes documentation in properties.md and man pages.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Add --uninitialized-check-all for address-taken variables

Extend uninitialized variable checking to also cover address-taken
(dirty) local variables using CBMC's shadow memory subsystem.

While reading uninitialized dirty locals is NOT undefined behavior
per C11 §6.3.2.1p2 (the value is merely indeterminate), it often
indicates a bug. This check is therefore separate from the C UB
check (--uninitialized-check).

The GOTO transformation pass (uninitialized_precise.cpp):
1. Adds shadow memory built-in symbols and empty function bodies
2. Inserts __CPROVER_field_decl_local in __CPROVER_initialize
3. After each write (direct or through pointer), inserts set_field
4. Before each read of a tracked dirty local, inserts get_field+assert
5. Skips address_of subtrees (taking address is not a read)

Symex resolves pointer aliasing and interprocedural writes precisely
via the value set, giving full precision at 0-20% overhead.

Includes documentation in properties.md and man pages, comparison
test suite, and regression tests for dirty variable patterns.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

kroening commented on doc/cprover-manual/properties.md:
```diff
@@ -129,7 +129,7 @@ The goto-instrument program supports these checks:
 | `--undefined-shift-check`    |  add range checks for shift distances                |
 | `--nan-check`                |  add floating-point NaN checks                       |
-| `--uninitialized-check`      |  add checks for uninitialized locals (experimental)  |
+| `--uninitialized-check`      |  add checks for reads of uninitialized locals (C11 UB) |
```
This makes it sound like uninitalized local reads have been UB only since C11 -- my understanding is that this was UB since C90.

kroening commented on doc/cprover-manual/properties.md:
```diff
@@ -207,6 +207,67 @@ unsigned foo(unsigned x)
+- Heap allocations (via pointer init flags)
+
+**What is not checked (by design, per C11 §6.3.2.1p2):**
+- Variables whose address has been taken ("dirty" variables) — reading
```
This should say "local variables". There is talk about heap allocated variables above.
